[MKM] Deadly Cover-Up, also genericize variants of ExileTargetAndSearchGraveyardHandLibraryEffect

ssk97 opened the pull request.

ssk97 commented:
Huh, so the only real difference between `controller.choose` and `controller.chooseTarget` is for tests, the former is set with `setChoice` and the latter with `addTarget`?

Since Surgical Extraction shouldn't target cards beyond the one initial target, modified the test to use `setChoice`.

JayDi85 commented:
It's important different cause "target" methods check target validation and
protection, so it's more restricted due rules. Non-target methods like
search allow much more possible targets.

In old days it was implemented by choose and chooseTarget. Now it uses
Target->notTarget field in most use cases. There are few exceptions but I
don't remember (must use setChoice instead addTarget in unit tests).

пт, 21 июн. 2024 г., 13:50 ssk97 ***@***.***>:

> Huh, so the only real difference between controller.choose and
> controller.chooseTarget is for tests, the former is set with setChoice
> and the latter with addTarget?
>
> Since Surgical Extraction shouldn't target cards beyond the one initial
> target, modified the test to use setChoice.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/magefree/mage/pull/12500#issuecomment-2182413294>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB7VEXIGEPIFH5LNAU42E7LZIPZNZAVCNFSM6AAAAABJVBM6HCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCOBSGQYTGMRZGQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

xenohedron commented on Mage/src/main/java/mage/abilities/effects/common/CounterTargetAndSearchGraveyardHandLibraryEffect.java:
```diff
@@ -20,13 +21,17 @@ public class CounterTargetAndSearchGraveyardHandLibraryEffect extends SearchTarg
     }
 
+    public CounterTargetAndSearchGraveyardHandLibraryEffect(boolean graveyardExileOptional, String searchWhatText, String searchForText, boolean drawForEachHandCard) {
+        super(graveyardExileOptional, searchWhatText, searchForText, drawForEachHandCard);
```
preferred if only one constructor calls the super

xenohedron commented on Mage/src/main/java/mage/abilities/effects/common/search/SearchTargetGraveyardHandLibraryForCardNameAndExileEffect.java:
```diff
@@ -59,37 +73,52 @@ protected boolean applySearchAndExile(Game game, Ability source, String cardName
+                cardsCount = Math.min(cardsInLibrary.count(filter, game), maxRemaining);
                 filter.setMessage("card named " + cardName + " in the library of " + targetPlayer.getLogName());
                 TargetCardInLibrary targetLib = new TargetCardInLibrary(0, cardsCount, filter);
+                target.withNotTarget(true);
```
TargetCardInLibrary already handles withNotTarget since it must always be true

xenohedron commented on Mage.Sets/src/mage/cards/n/Necromentia.java:
```diff
@@ -50,6 +50,7 @@ public Necromentia copy() {
     }
 }
 
+//Based on SearchTargetGraveyardHandLibraryForCardNameAndExileEffect
```
is it possible to make an overrideable method that applies an effect based on the value? not required.

ssk97 replied:
That could work, but right now this is the only card that does anything other than drawing so I don't think it's worth the complexity.

xenohedron reviewed (commented):
Thanks for the refactor.

Is there existing test coverage here? If not, could you please write some unit tests to confirm the functionality and the constructor parameters?

ssk97 commented:
> Is there existing test coverage here? If not, could you please write some unit tests to confirm the functionality and the constructor parameters?

There is the Surgical Extraction test but AFAIK nothing aside from that. I'll add a test for the card-draw count, not sure how to test for "up to 4" though, any suggestions there?

ssk97 commented:
While adding tests I noticed that we don't properly follow the name rules for DFC cards, xmage was matching against the back side when it shouldn't. Fixed that as well in the tests.

I did have to add a non-strict test since AFAIK there's no way to currently test for the absence of choice targets. I also added a 2nd non-strict test to confirm the expected behavior if the choices were available.
